runtime/src/lib.rs

```diff
@@ -656,74 +656,47 @@ impl_runtime_apis! {
 			moonbeam_rpc_primitives_debug::single::TransactionTrace,
 			sp_runtime::DispatchError
 		> {
-			// Get the caller;
-			let mut sig = [0u8; 65];
-			let mut msg = [0u8; 32];
-			sig[0..32].copy_from_slice(&transaction.signature.r()[..]);
-			sig[32..64].copy_from_slice(&transaction.signature.s()[..]);
-			sig[64] = transaction.signature.standard_v();
-			msg.copy_from_slice(
-				&pallet_ethereum::TransactionMessage::from(transaction.clone()).hash()[..]
-			);
-
-			let from = match sp_io::crypto::secp256k1_ecdsa_recover(&sig, &msg) {
-				Ok(pk) => H160::from(
-					H256::from_slice(Keccak256::digest(&pk).as_slice())
-				),
-				_ => H160::default()
-			};
+			use moonbeam_rpc_primitives_debug::single::TraceType;
 
 			// Apply the a subset of extrinsics: all the substrate-specific or ethereum transactions
 			// that preceded the requested transaction.
 			for ext in extrinsics.into_iter() {
 				let _ = match &ext.function {
 					Call::Ethereum(transact(t)) => {
 						if t == transaction {
-							break;
+							let hook = match trace_type {
+								TraceType::Raw {
+									disable_storage,
+									disable_memory,
+									disable_stack,
+								} => {									
+									TracingHook::new_raw(
+										disable_storage,
+										disable_memory,
+										disable_stack,
+									)									
+								},
+								TraceType::CallList => todo!(),
+							};
+
+							let other_hook = pallet_evm::runner::stack::Runner::<Runtime, TracingHook>::set_hook(Some(hook));		
+							let _ = Executive::apply_extrinsic(ext);
```

Comment:
This here is mind blowing, really smart.

To make sure I understand this, what you are doing is:

- Writting a `Hook` implementor in a static map.
- Then at runner level you swap the hook with `None`, that will give you back the hook you are setting here (or `None` in a non-tracing EVM execution, because we are not setting any hook).
- Apply extrinsic will call the evm with the hook enabled, so it will `execute` in the EVM executor with stepping enabled using the custom `Hook`'s implementation.
- You get the hook back, you call finish and return the trace.

Reply:
Exactly. At runner level we get back the hook from the EVM to put it back in the static.